### ISSUE TITLE
OE-14 "Adding FP16 path in DNN" discussion

### ISSUE THREAD

**Issue by vpisarev** [opencv/opencv#11009]
the feature request about evolution proposal [OE-14](https://github.com/opencv/opencv/wiki/OE-14.-DNN-FP16)

**Comment by hrnr** [comment]
What about int8 compute path in DNN? Quantized networks are already supported for importing.

int8 path possibly offers a huge speedup and may be also implemented for CPU. Huge speedup, memory savings and availability on CPU makes int8 path even more appealing to me than FP16. 

ref: https://www.tensorflow.org/performance/quantization

**Comment by pengli** [comment]
I have some patches available for dnn ocl fp16 support in my branch at https://github.com/pengli/opencv/tree/dnn_half , currently only convolution half support is added.

**Comment by vpisarev** [comment]
@hrnr, I agree that in more distant future int8 has more potential. But it's more difficult to do "right". In the case of FP16 we can simply convert both weights _and_ the intermediate activations from FP32 to FP16, without any fine-tuning, and hopefully get very similar results; many modern GPUs support FP16 out of the box and we can expect nearly 2x better performance. Right now. In the case of int8 we can get better performance on CPUs (very few GPUs have dedicated INT8 ALUs nowadays), but in order to get that, we need to quantise activations, and that's very difficult to do well if we have just the model, without the training dataset.

So, it's "2x speedup right now, without any problems" vs "possibly 2x speedup (some experiments by my colleagues indicate it's more modest acceleration in practice) with more hassle and higher risk of loosing accuracy"

**Comment by vpisarev** [comment]
@pengli, yes, I remember our brief conversation. I hope, we will eventually make some use of your patches ;)

**Comment by pengli** [comment]
@vpisarev,  these patches are quick prototype to test fp16 ocl support in dnn. I add some hack in the framework to make it support fp16.  so far it need "convert float to half" for the input blob at the layer forward begin and "convert half to float" for the output blob at the layer forward end. That is not efficient.  Eventually they should be removed by using pure fp16  data in the layer forward.

**Comment by alalek** [comment]
BTW, valgrind (3.13.0 from Fedora 26 or Ubuntu) doesn't work well with x86 FP16 instructions:
<details>

```
[ RUN      ] Test_TensorFlow.fp16
vex amd64->IR: unhandled instruction bytes: 0xC4 0xC2 0x7D 0x13 0x44 0x55 0x0 0xC4 0xC1 0x78
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=1
vex amd64->IR:   VEX=1 VEX.L=1 VEX.nVVVV=0x0 ESC=0F38
vex amd64->IR:   PFX.66=1 PFX.F2=0 PFX.F3=0
==15851== valgrind: Unrecognised instruction at address 0x949015e.
==15851==    at 0x949015E: _mm256_cvtph_ps (f16cintrin.h:54)
==15851==    by 0x949015E: cv::opt_FP16::cvtScaleHalf_SIMD16f32f(short const*, unsigned long, float*, unsigned long, cv::Size_<int>) (convert.fp16.cpp:57)
==15851==    by 0x92A9818: cvtScaleHalf_<short int, float> (convert.cpp:1383)
==15851==    by 0x92A9818: cv::cvtScaleHalf16f32f(short const*, unsigned long, float*, unsigned long, cv::Size_<int>) (convert.cpp:1407)
==15851==    by 0x92B303D: cv::convertFp16(cv::_InputArray const&, cv::_OutputArray const&) (convert.cpp:1488)
```

</details>

**Comment by pengli** [comment]
@alalek , will there be a CPU version fp16 implementation in layer forwarding ?

**Comment by alalek** [comment]
These instructions are used to convert `fp16` <-> `fp32`.
FP16 DNN CPU implementation make sense to reduce used memory bandwidth (load fp16, convert and calculate in fp32, convert back and store fp16 result). We thinking about this, but currently there are no details.

**Comment by dkurt** [comment]
In the PR https://github.com/opencv/opencv/pull/11062 introduced FP16 GPU support for clDNN plugin of Inference Engine backend. Unfortunately, only a couple of models are supported by current DLSDK release. This is one of them:

| | Default backend, CPU | IE backend, CPU | Default backend, GPU, FP32 | IE backend, GPU, FP32 |
|----|---|---|---|---|
| OpenPose, pose estimation, COCO |  879.13ms | 543.942ms | 1048.94ms | 758.36ms |
| IE backend, GPU, FP16<br>**466.48ms** | x1.88  | x1.16 | x2.24| x1.62 |

**Comment by pengli** [comment]
https://github.com/pengli/opencv/tree/dnn_half  is updated, it can now support fp16 in googlenet and alexnet. but it requires fp32->fp16 at layer forward begin and fp16->fp32 at layer forward end.

**Comment by pengli** [comment]
fyi, https://github.com/pengli/opencv/tree/dnn_half is updated,  it adds support for CV_16S format in the layer allocation, so it allows fp16 input/output blob in the layer forward.  For layers that don't support fp16, half<->float conversion is required, otherwise the conversion can be avoided.

**Comment by borisfom** [comment]
@vpisarev : any plans to add CUDA path to DNN? 
In the meantime, I have prepared this PR to introduce CV_16F depth as first-class citizen: 
https://github.com/opencv/opencv/pull/10485

**Comment by dkurt** [comment]
It looks like we can close this evolution proposal as done.
